core/lib/zksync_core/src/api_server/web3/namespaces/eth.rs

```diff
@@ -490,6 +491,24 @@ impl EthNamespace {
                 .map_err(|err| internal_error(method_name, err))?;
         }
 
+        // Number of txs accepted by the node.
+        // We should add the nonces from txs in our cache for making the nonce
+        // aligned with the main node. If the nonce is not sequential,
+        // then we should not increment nonce.
```

Comment:
But TxCache lives only in memory and we do restart for all reverts, so it will be empty after the revert. 

For now we have either mempool or tx cache (mainnode vs en) so they don't work together

Reply:
> so it will be empty after the revert

I get this, but the EN will still report transaction count *including* the reverted transactions, as if they are present in the mempool. Maybe this is a feature, not a bug, I just want to highlight that it's inconsistent with the "ordinary" EN restart (after which the EN forgets all transactions in the mempool).

> For now we have either mempool or tx cache (mainnode vs en) so they don't work together

This is non-local reasoning; I feel that it should be explained as a comment. Also, it could be easily avoided if the code is refactored as follows:

```rust
if matches!(block_id, BlockId::Number(BlockNumber::Pending)) {
    let pending_nonces: Vec<Nonce> = if let Some(proxy) = &self.state.tx_sender.0.proxy {
        // EN: get pending nonces from the transaction cache
        proxy.get_nonces_by_account(address).await
    } else {
        // Main node: load pending nonces from the database
        connection
            .transactions_web3_dal()
            .pending_nonces_by_initiator_account(start_nonce) // minor refactoring of `next_nonce_by_initiator_account`
            .await
            .map_err(|err| internal_error(method_name, err))?
    };
    // find out the next free nonce based on `pending_nonces`.
}
```

Maybe, it makes sense instead to refactor `get_nonces_by_account` so that it has the same semantics as `pending_nonces_by_initiator_account` (i.e., accepts a starting nonce and returns the next free nonce). It could be beneficial because it makes it easier to detect a currently present bug: `get_nonces_by_account` may return outdated nonces (i.e., ones lesser than `actual_tx_number`) because they are updated asynchronously. In this case, the current logic won't work correctly.

Also, you haven't addressed the concern with non-pending block IDs. Unless I'm missing something obvious, pending nonces should only be taken into account for `BlockNumber::Pending`, but currently they are taken into account for all requests on the EN. This can lead to bogus values returned (e.g., for `BlockNumber::Latest`).

Reply:
Finally, i get it. 
Yes you are right. 
We have mempool in main node and we have tx proxy in EN, we have to use tx proxy as mempool. 

It doesn't make sense to get pending nonce from the db for EN, because it's always empty